File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr):
        # Step 1: sort unique elements
        sorted_unique = sorted(set(arr))
        
        # Step 2: assign ranks
        rank_map = {num: i+1 for i, num in enumerate(sorted_unique)}
        
        # Step 3: build result
        return [rank_map[num] for num in arr]

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        vector<int> sortedArr = arr;
        sort(sortedArr.begin(), sortedArr.end());

        unordered_map<int,int> rank;
        int r = 1;
        for (int x : sortedArr) {
            if (rank.find(x) == rank.end()) {
                rank[x] = r++;
            }
        }

        vector<int> result;
        for (int x : arr) {
            result.push_back(rank[x]);
        }
        return result;
    }
};
